src/core_scan/Report_pro_findings.ml

```diff
@@ -0,0 +1,107 @@
+(*****************************************************************************)
+(* Purpose *)
+(*****************************************************************************)
+(** Support the reporting of pro findings
+
+  * Users want to know when a finding could only be found by the pro engine
+  * so that they can understand the value Semgrep provides. Additionally,
+  * we want to surface our "coolest" findings early to make sure users see
+  * them.
+
+  * To support this, we have a different type for the engine kind for matches,
+  * which includes a `PRO_SPECIFIC` option that reports when a finding could
+  * only be found using the pro engine. This option also allows us to highlight
+  * findings we think users should prioritize.
+  *)
+
+module FileSet = Set.Make (String)
+
+(*****************************************************************************)
+(* Helpers *)
+(*****************************************************************************)
+
+(* Check if the taint trace of the pattern crosses files *)
+
+let union3 a b c = FileSet.union a (FileSet.union b c)
+
+let files_of_toks (ts : Pattern_match.pattern_match_tokens) : FileSet.t =
+  ts
+  |> List.fold_left
+       (fun acc_files t ->
+         match Tok.loc_of_tok t with
+         | Ok { pos; _ } -> FileSet.add pos.file acc_files
+         | Error _ -> acc_files)
+       FileSet.empty
+
+let files_of_trace_item (trace_item : Pattern_match.taint_trace_item) =
+  let rec files_of_call_trace (call_trace : Pattern_match.taint_call_trace) =
+    match call_trace with
+    | Toks ts -> files_of_toks ts
+    | Call { call_toks; intermediate_vars; call_trace } ->
+        let intermediate_var =
+          match intermediate_vars with
+          | [] -> []
+          | x :: _ -> [ x ]
```

Comment:
Actually I think you don't even need to look at any `intermediate_vars`. If you look at the source, the sink, and the call (toks), the intermediate vars should not be adding anything new.

Reply:
That makes sense; thanks for clarifying